src/pathToVictory/services/officeMatch.service.ts

```diff
@@ -24,31 +27,197 @@ export class OfficeMatchService {
+        officeName,
+      )
+      heuristicLevelTypes = base.filter((t) => typesSet.has(t))
+    } catch (e) {
+      const msg = `Error determining heuristic level types: ${e instanceof Error ? e.message : String(e)}`
+      this.logger.error(msg)
+    }
+    this.logger.debug(
+      `searchDistrictTypes: heuristicLevelTypes=${JSON.stringify(heuristicLevelTypes)}`,
+    )
+
+    // 2b) Sub-district types if we likely have a numeric district (filtered to API types)
+    let subColumns: string[] = []
+    const districtValue = this.getDistrictValue(
+      slug,
+      officeName,
+      subAreaName,
+      subAreaValue,
+    )
+    if (
+      electionLevel !== 'federal' &&
+      electionLevel !== 'state' &&
+      heuristicLevelTypes.length > 0 &&
+      districtValue
+    ) {
+      const subs = this.determineElectionDistricts(
+        slug,
+        heuristicLevelTypes,
+        officeName,
+      )
+      subColumns = subs.filter((t) => typesSet.has(t))
+    }
+    this.logger.debug(
+      `searchDistrictTypes: subColumns=${JSON.stringify(subColumns)} (districtValue=${districtValue})`,
+    )
+
+    // 2c) Category-filtered types (e.g., City/County/State/Judicial/Education)
+    let categoryTypes: string[] = districtTypes
+    if (category) {
+      const cat = category.toLowerCase()
+      const filtered = districtTypes.filter((t) =>
+        String(t).toLowerCase().includes(cat),
       )
+      if (filtered.length > 0) {
+        categoryTypes = filtered
+      }
+    }
+    this.logger.debug(
+      `searchDistrictTypes: categoryTypes (post-filter) size=${categoryTypes.length}`,
+    )
 
-      this.logger.debug('miscDistricts', searchColumns)
-      return searchColumns
-    } catch (error) {
-      this.logger.error('error', error)
+    // 2d) Combine candidates in priority order: subColumns → heuristicLevelTypes → categoryTypes → remaining
+    const orderedCandidates: string[] = []
+    const pushUnique = (arr: string[]) => {
+      for (const t of arr) {
+        if (typesSet.has(t) && !orderedCandidates.includes(t)) {
+          orderedCandidates.push(t)
+        }
+      }
     }
-    return searchColumns
+    pushUnique(subColumns)
+    pushUnique(heuristicLevelTypes)
+    pushUnique(categoryTypes)
+    // add remaining API types not yet included
+    pushUnique(districtTypes)
+    this.logger.debug(
+      `searchDistrictTypes: orderedCandidates size=${orderedCandidates.length}`,
+    )
+
+    // 3) Use AI to select the top 1-5 best matching district type columns
+    const matchResp = await this.matchSearchColumns(
+      slug,
+      orderedCandidates,
+      officeName,
+    )
+
+    let foundDistrictTypes: string[] = []
+    if (matchResp?.content) {
+      try {
+        const parsed: unknown = JSON.parse(matchResp.content)
+        if (
+          parsed &&
+          typeof parsed === 'object' &&
+          'columns' in parsed &&
+          Array.isArray((parsed as { columns?: unknown[] }).columns)
```

Comment:
This seems extremely verbose.. why not just do an existential check on `parsed.columns`? Why the magic string object key check? 🤔